### PR TITLE
[WFE-303] Rewrite Markdown extensions to work in webpack

### PR DESCRIPTION
**Pull request by chudoklates**
### Description
Rewrite all extensions to pass the assertions set in development mode of `micromark`

### Manual PR Checklist

- [x] CHANGELOG entry added
- [x] Test coverage added
- [x] Licenses checked
- [x] All related Jira tickets added to PR title
- [x] Description in Jira ticket is clear and up-to-date

[WFE-303]: https://labforward.atlassian.net/browse/WFE-303?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bin/lib.mjs [hunk 1]
```diff
@@ -6,6 +6,22 @@ const raw = `
   ***bold***
   1. list 1
   2. list 2
+
+  # TEST INTERPOLATION
```
Comment (chudoklates) [inline bin/lib.mjs:10]: We can write an additional test case for checks like this, for now I put it here not to mess with the snapshot.

File: src/extensions/interpolation/syntax.ts [hunk 2]
```diff
@@ -78,19 +91,19 @@ function tokenizeInterpolation(effects: Effects, ok: State, nok: State) {
...
   }
 
   function onInterpolationEnd(code: Code) {
+    // if the final code is '}', interpolation is complete,
+    // if it is anything else, we abort with nok
     if (code === codes.rightCurlyBrace) {
       effects.consume(code);
-      markers -= 1;
-
-      if (markers) return onInterpolationEnd;
```
Comment (ignatiusreza) [inline src/extensions/interpolation/syntax.ts:89]: why are these removed?

this was previously used to ensure that the interpolation ended with the exact number of `}` (2 characters)

File: bin/lib.mjs (hunk 1 above)
Comment (ignatiusreza) [reply bin/lib.mjs:10]: I agree.. lets write more test cases :+1:

File: src/extensions/interpolation/syntax.ts [hunk 3]
```diff
@@ -35,21 +38,31 @@ function tokenizeInterpolation(effects: Effects, ok: State, nok: State) {
...
+      }
 
-      return onInterpolationFormula;
+      return onInterpolationStart;
     }
 
-    return nok;
+    // return nok(code) instead of nok to ensure the wrong
+    // character is also consumed
+    return nok(code);
```
[the remark this answers is not in the document]
Comment (chudoklates) [reply src/extensions/interpolation/syntax.ts:56]: it's what micromark expects. I realised this mostly through trial & error, but without consuming the wrong character we fail the built-in assertions. I was also referring to [this section of their README](https://github.com/micromark/micromark/tree/main#creating-a-micromark-extension), and there the `nok` is also called with the code.

[the remark this answers is not in the document]
Comment (chudoklates) [reply src/extensions/interpolation/syntax.ts:56]: I think the general idea is that every character must be consumed, no matter if it fits the extension's syntax or not.

File: src/extensions/interpolation/syntax.ts (hunk 2 above)
Comment (chudoklates) [reply src/extensions/interpolation/syntax.ts:89]: I removed it because this is now done implicitly.

```js
  function onInterpolationFormula(code) {
    if (code === codes.rightCurlyBrace) {
      effects.exit(type);
      effects.enter("void");
      effects.consume(code);

      return onInterpolationEnd;
    }
```

the first `}` is already consumed by the time we call `onInterpolationEnd`. Therefore if we encounter this character within `onInterpolationEnd`, it is already the second instance, and if we encounter anything else (like in `{{f}x`) we should terminate. Beforehand, we would return `onInterpolationEnd` from the aforementioned if block without consuming, but this is also not allowed by micromark.

**Comment by chudoklates** [comment]
I will use this branch in WFE before merging it to verify if everything is indeed working fine. In the meantime, feel free to give approval @ignatiusreza

**Comment by chudoklates** [comment]
> I will use this branch in WFE before merging it to verify if everything is indeed working fine. In the meantime, feel free to give approval @ignatiusreza

All good to go ✅

File: bin/lib.mjs (hunk 1 above)
Comment (chudoklates) [reply bin/lib.mjs]: https://labforward.atlassian.net/browse/WFE-311 created a ticket for this so we can keep track. You can add more test cases here if you can think of any.

**Review by ignatiusreza — APPROVED** [review]
LGTM :+1:

#### Review checklist according to [SOP-13][1] & [SOP-014][2]

- [x] Code is readable
- [x] Test coverage is appropriate
- [x] CHANGELOG entry is correct
- [x] I manually tested the changes and checked that related capabilities do not have any regressions
- [x] I confirmed that all specifications linked to the issue are fulfilled
- [x] I checked newly added dependencies for their licenses

[1]: https://drive.google.com/file/d/1fD1ZcsBTywlMVTObTmNLrg55khrYMI_L/view
[2]: https://drive.google.com/file/d/1uRnMtykgw1Gzo2nRp-TPnESgmYoEIfBf/view
